table/block_based_table_reader.cc

```diff
@@ -1759,30 +1764,37 @@ BlockBasedTable::PartitionedIndexIteratorState::NewSecondaryIterator(
 //
 // REQUIRES: this method shouldn't be called while the DB lock is held.
 bool BlockBasedTable::PrefixMayMatch(const Slice& internal_key,
-                                     const SliceTransform* prefix_extractor) {
+                                     const ReadOptions& read_options,
+                                     const bool prefix_extractor_changed) {
   if (!rep_->filter_policy) {
     return true;
   }
 
-  assert(prefix_extractor != nullptr);
+  if (rep_->table_prefix_extractor == nullptr) {
+    return true;
+  }
   auto user_key = ExtractUserKey(internal_key);
-  if (!prefix_extractor->InDomain(user_key)) {
+  if (!rep_->table_prefix_extractor->InDomain(user_key)) {
     return true;
   }
-  assert(rep_->table_properties->prefix_extractor_name.compare(
-             prefix_extractor->Name()) == 0);
-  auto prefix = prefix_extractor->Transform(user_key);
 
   bool may_match = true;
   Status s;
 
   // First, try check with full filter
-  auto filter_entry = GetFilter(prefix_extractor);
+  auto filter_entry = GetFilter(rep_->table_prefix_extractor.get());
   FilterBlockReader* filter = filter_entry.value;
+  // set prefix using prefix_extractor from the SST
+  auto prefix = rep_->table_prefix_extractor->Transform(user_key);
   if (filter != nullptr) {
     if (!filter->IsBlockBased()) {
       const Slice* const const_ikey_ptr = &internal_key;
-      may_match = filter->PrefixMayMatch(prefix, prefix_extractor, kNotValid,
+      if (prefix_extractor_changed &&
+          !filter->IsFilterCompatible(read_options, user_key,
+                                      rep_->table_prefix_extractor.get())) {
```

Comment:
Rather than calling IsFilterCompatible(). Please consolidate IsFilterCompatible() and PrefixMayMatch() into one function, which may call something like ExistsInRange().